app/features/worksheet_generator/tools.py

```diff
@@ -0,0 +1,481 @@
+                logger.info(f"Generated response attempt {attempts + 1}: {response}")
+
+            if response is None:
+                if self.verbose:
+                    logger.warning(f"Invalid null response. Attempt {attempts + 1} of {max_attempts}")
+                attempts += 1
+                continue
+
+            response = self.transform_json_dict(response)
+
+            # Directly check if the response format is valid
+            if self.validate_response(response):
+                response["choices"] = self.format_choices(response["choices"])
+                generated_questions.append(response)
+                if self.verbose:
+                    logger.info(f"Valid question added: {response}")
+                    logger.info(f"Total generated questions: {len(generated_questions)}")
+            else:
+                if self.verbose:
+                    logger.warning(f"Invalid response format. Attempt {attempts + 1} of {max_attempts}")
+            
+            # Move to the next attempt regardless of success to ensure progress
+            attempts += 1
+
+        # Log if fewer questions are generated
+        if len(generated_questions) < num_questions:
+            logger.warning(f"Only generated {len(generated_questions)} out of {num_questions} requested questions")
+        
+        # Return the list of questions
+        return generated_questions[:num_questions]
+    
+class FillInBlankQuestionBuilder(QuestionBuilder):
+    def __init__(self, grade_level, topic, difficulty_level, lang, prompt=None, model=None, parser=None, verbose=False):
+        default_config = {
+            "model": GoogleGenerativeAI(model="gemini-1.0-pro"),
+            "parser": JsonOutputParser(pydantic_object=FillInBlankQuestion),
+            "prompt": read_text_file("prompt/fill-in-blank-question-builder-prompt.txt")
+        }
+        
+        self.prompt = prompt or default_config["prompt"]
+        self.model = model or default_config["model"]
+        self.parser = parser or default_config["parser"]
+        
+        self.grade_level = grade_level
+        self.topic = topic
+        self.difficulty_level = difficulty_level
+        self.lang = lang
+        self.verbose = verbose
+        
+        if grade_level is None: raise ValueError("Grade level must be provided")
+        if topic is None: raise ValueError("Topic must be provided")
+        if difficulty_level is None: raise ValueError("Difficulty level must be provided")
+
+    def transform_json_dict(self, input_data: Dict) -> Dict:
+        # Validate and parse the input data to ensure it matches the FillInBlankQuestion schema
+        quiz_question = FillInBlankQuestion(**input_data)
+
+        # Create the transformed structure
+        transformed_data = {
+            "question": quiz_question.question,
+            "answer": quiz_question.answer,
+        }
+
+        return transformed_data
+    
+    def validate_response(self, response: Dict) -> bool:
```

Comment:
This validation is not required if the use of the JSONOutputParser is effectively done with the given prompt.